peppy/project.py

```diff
@@ -140,16 +150,146 @@ def __init__(
             index=self.st_index, initial=True
         )
 
-    def to_dict(self, expand=False):
+        # init project from dict
+        if project_dict:
+            self.from_dict(project_dict)
+
+    def __eq__(self, other):
+
+        dict_self = self._convert_to_dict(self)
+        dict_other = self._convert_to_dict(other)
+
+        dict_self["_samples"] = [s.to_dict() for s in self.samples]
+        dict_other["_samples"] = [s.to_dict() for s in other.samples]
+
+        return dict_self == dict_other
+
+    def _convert_to_dict(self, project_value=None):
+        """
+        Recursively transform various project values, objects, and attributes to a dictionary
+        compatible format. Useful for creating an extended dictionary representation
+        of the peppy project.
+
+        :param project_value object - the value to transform
+        """
+        if isinstance(project_value, list):
+            new_list = []
+            for item_value in project_value:
+                new_list.append(self._convert_to_dict(item_value))
+            return new_list
+
+        elif isinstance(project_value, dict):
+            new_dict = {}
+            for key, value in project_value.items():
+                if key != "_project":
+                    new_dict[key] = self._convert_to_dict(value)
+            return new_dict
+
+        elif isinstance(project_value, PathExAttMap):
+            new_dict = PathExAttMap.to_dict(project_value)
+            return self._convert_to_dict(new_dict)
+            # return new_dict
+
+        elif isinstance(project_value, Sample):
+            new_dict = PathExAttMap.to_dict(project_value)
+            return new_dict
+
+        elif isinstance(project_value, pd.DataFrame):
+            project_value = project_value.to_dict()
+            project_value = self._nan_converter(project_value)
+            return project_value
+
+        else:
+            return project_value
+
+    def _nan_converter(self, nan_dict: dict) -> dict or list:
+        """
+        Searching and converting nan values to None
+        :param dict nan_dict: dictionary with nan values
+        """
+        if isinstance(nan_dict, list):
+            new_list = []
+            for list_item in nan_dict:
+                new_list.append(self._nan_converter(list_item))
+
+            return new_list
+
+        elif isinstance(nan_dict, dict):
+            new_dict = {}
+            for key, value in nan_dict.items():
+                new_dict[key] = self._nan_converter(value)
+            return new_dict
+        elif isinstance(nan_dict, float):
+            if math.isnan(nan_dict):
+                return None
+        else:
+            return nan_dict
+
+    def from_dict(self, d: dict) -> None:
```

Comment:
Maybe replace `d` variable with something more verbose? Eg. `dictionary_representation`, `pep_dictionary_representation`?

Reply:
I think I wrote this function, and I agree. I'm partial to `pep_dictionary`